File: app/src/main/java/com/example/review/SortLib.java
package com.example.review;

import com.example.review.New.ArrayStoreList;

import java.io.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

public class SortLib extends ArrayStoreList<SortLib.SortStru> {
    public SortLib() {

    }

    public SortLib(String src) {
        decode(src);
    }

    public SortLib(File src) {
        try {
            read(src);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void decode(String src) {
        String[] sorts = src.split("\n");
        for (String sortRaw : sorts) {
            if (!sortRaw.isEmpty()) add(new SortStru(sortRaw));
        }
    }

    public List<String> getElementInSort(String sort) {
        for (SortStru item : this) {
            if (item.matchSort(sort)) {
                //防止多次引用
                if (item.inUse) return null;
                item.inUse = true;

                List<String>      elements = item.getElements();
                ArrayList<String> temp     = new ArrayList<>();
                ArrayList<String> refer    = new ArrayList<>();

                //把引用的元素列举出来
                for (String element : elements) {
                    if (element.contains("@")) {
                        refer.add(element);
                        element = element.substring(1);
                        List<String> elementInSort = getElementInSort(element);
                        temp.addAll(elementInSort);
                    }
                }
                elements.addAll(temp);
                elements.removeAll(refer);
                item.inUse = false;
                return elements;
            }
        }
        return new ArrayList<>();
    }

    @Override
    public String toString() {
        return toString(false);
    }

    public String toString(boolean listAll) {
        StringBuilder sb = new StringBuilder();
        for (SortStru sortStru : this) sb.append("\n" + sortStru.toString(listAll ? this : null));
        sb.replace(0, 1, "");
        return sb.toString();
    }

    class SortStru {
        private String       name;
        private List<String> elements = new LinkedList<>();
        private boolean      inUse;

        SortStru(String src) {
            String[] split = src.split("[：:]");
            name = split[0];
            if (split.length == 2) {
                String[] elements = split[1].split("[、,]");
                this.elements.addAll(Arrays.asList(elements));
            }
        }

        SortStru(String name, List<String> elements) {
            setName(name);
            setElements(elements);
        }

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }

        public void setElements(List<String> elements) {
            this.elements = elements;
        }

        public List<String> getElements() {
            return new ArrayList<>(elements);
        }

        public boolean matchSort(String sortName) {
            return sortName.equals(name);
        }

        public List<String> getElements(SortLib sortStrus) {
            return sortStrus.getElementInSort(name);
        }

        @Override
        public String toString() {
            return toString(null);
        }

        public String toString(SortLib sortStrus) {
            StringBuffer sb       = new StringBuffer();
            List<String> elements = (sortStrus != null) ? getElements(sortStrus) : getElements();
            for (String element : elements) sb.append(element).append('、');
            int length = sb.length();
            sb.replace(length - 1, length, "");
            String format = String.format("%s：%s", name, sb);
            return format;
        }
    }

    @Override
    public void toBytes(DataOutputStream dos) throws IOException {
        StringBuffer sb = new StringBuffer();
        for (SortStru sortStru : this) {
            String str = sortStru.toString();
            sb.append("\n").append(str);
        }
        sb.replace(0, 1, "");
        String value = sb.toString();
        dos.write(value.getBytes());
    }

    @Override
    public void loadWith(DataInputStream dis) throws IOException {
        byte[]                buffer = new byte[1024];
        ByteArrayOutputStream baos   = new ByteArrayOutputStream();
        int                   length;
        while ((length = dis.read(buffer)) != -1) baos.write(buffer, 0, length);
        decode(new String(baos.toByteArray()));
    }
}

File: app/src/main/java/com/example/review/SortLib.kt
package com.example.review

import com.example.review.New.ArrayStoreList
import com.example.review.SortLib.SortStru
import java.io.*
import java.util.*
import kotlin.collections.ArrayList

class SortLib : ArrayStoreList<SortStru> {
    constructor()

    constructor(src: String) {
        decode(src)
    }

    constructor(src: File) {
        try {
            read(src)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    private fun decode(src: String) {
        val sorts = src.split("\n").toTypedArray()
        for (sortRaw in sorts) {
            if (!sortRaw.isEmpty()) add(SortStru(sortRaw))
        }
    }

    fun getElementInSort(sort: String): List<String> {
        for (item in this) {
            if (item.matchSort(sort)) { //防止多次引用
                if (item.inUse) return ArrayList()
                item.inUse = true
                val elements = item.getElements()
                val temp = ArrayList<String>()
                val refer = ArrayList<String>()
                //把引用的元素列举出来
                for (element in elements) {
                    var element = element
                    if (element.contains("@")) {
                        refer.add(element)
                        element = element.substring(1)
                        val elementInSort = getElementInSort(element)
                        temp.addAll(elementInSort!!)
                    }
                }
                elements.addAll(temp)
                elements.removeAll(refer)
                item.inUse = false
                return elements
            }
        }
        return ArrayList()
    }

    override fun toString(): String {
        return toString(false)
    }

    fun toString(listAll: Boolean): String {
        val sb = StringBuilder()
        for (sortStru in this) sb.append("\n" + sortStru.toString(if (listAll) this else null))
        sb.replace(0, 1, "")
        return sb.toString()
    }

    inner class SortStru {
        var name: String = ""
        private var elements: MutableList<String> = LinkedList()
        var inUse = false

        constructor(src: String) {
            val split = src.split("[：:]").toTypedArray()
            name = split[0]
            if (split.size == 2) {
                val elements = split[1].split("[、,]").toTypedArray()
                this.elements.addAll(Arrays.asList(*elements))
            }
        }

        constructor(name: String, elements: MutableList<String>) {
            this.name = name
            setElements(elements)
        }

        fun setElements(elements: MutableList<String>) {
            this.elements = elements
        }

        fun getElements(): MutableList<String> {
            return ArrayList(elements)
        }

        fun matchSort(sortName: String?): Boolean {
            return sortName == name
        }

        fun getElements(sortStrus: SortLib): List<String>? {
            return sortStrus.getElementInSort(name)
        }

        override fun toString(): String {
            return toString(null)
        }

        fun toString(sortStrus: SortLib?): String {
            val sb = StringBuffer()
            val elements = sortStrus?.let { getElements(it) } ?: getElements()
            for (element in elements) sb.append(element).append('、')
            val length = sb.length
            sb.replace(length - 1, length, "")
            return String.format("%s：%s", name, sb)
        }
    }

    @Throws(IOException::class)
    override fun toBytes(dos: DataOutputStream) {
        val sb = StringBuffer()
        for (sortStru in this) {
            val str = sortStru.toString()
            sb.append("\n").append(str)
        }
        sb.replace(0, 1, "")
        val value = sb.toString()
        dos.write(value.toByteArray())
    }

    @Throws(IOException::class)
    override fun loadWith(dis: DataInputStream) {
        val buffer = ByteArray(1024)
        val baos = ByteArrayOutputStream()
        var length: Int
        while (dis.read(buffer).also { length = it } != -1) baos.write(buffer, 0, length)
        decode(String(baos.toByteArray()))
    }
}